internal/pkg/graph/graph.go

```diff
@@ -100,21 +123,65 @@ func (g *Graph[V]) hasCycles(temp *findCycleTempVars[V], currVertex V) bool {
 	return false
 }
 
-// TopologicalSorter ranks vertices in a graph using topological sort.
-type TopologicalSorter[V comparable] struct {
-	ranks map[V]int
+// LevelOrder ranks vertices in a breadth-first search manner.
+type LevelOrder[V comparable] struct {
+	marked map[V]bool
+	ranks  map[V]int
 }
 
 // Rank returns the order of the vertex. The smallest order starts at 0.
-// If the vertex does not exist in the graph, then returns an error.
-func (s *TopologicalSorter[V]) Rank(vtx V) (int, error) {
-	// TODO(efekarakus): Implement me.
-	return 0, nil
+// The second boolean return value is used to indicate whether the vertex exists in the graph.
+func (bfs *LevelOrder[V]) Rank(vtx V) (int, bool) {
+	r, ok := bfs.ranks[vtx]
+	return r, ok
+}
+
+func (bfs *LevelOrder[V]) traverse(g *Graph[V], root V) {
+	bfs.marked = make(map[V]bool) // Reset all markings before each run.
+
+	queue := []V{root}
+	bfs.marked[root] = true
+	bfs.ranks[root] = 0
+	for len(queue) > 0 {
+		var vtx V
+		vtx, queue = queue[0], queue[1:]
+		for neighbor := range g.vertices[vtx] {
+			if bfs.marked[neighbor] {
+				continue
+			}
+			bfs.marked[neighbor] = true
+			if rank := bfs.ranks[vtx] + 1; rank > bfs.ranks[neighbor] { // Is the new rank higher than a previous traversal?
```

Comment:
I might be stupid but I am a little confused here 🥲 Only when `bfs.marked[neighbor]` is false can the program proceed here, but then does that mean `bfs.ranks[neighbor]` will be empty too?

I feel like if a vertex is visited, we shouldn't skip too quickly but possibly think about if we should update the rank and update with the lowest rank rather than the highest. For example:
```
a -> b -> c
d -> c -> f
```
where both `a` and `d` has 0 rank. If we give `c` rank 2 rather than 1 then `f` will share the same rank as `c`.

Reply:
lol it's not stupid - 

> but then does that mean bfs.ranks[neighbor] will be empty too?

I think your example shows how it's possible that `ranks` is not empty. If your example, after `a -> b -> c` we will have:
```go
map[string]int {
    "a": 0,
    "b": 1,
    "c": 2
}
```
Then when we start from the second root, `d -> c -> f` we **do not** want the rank of `c` to become 1. It has a higher rank from a previous deployment, so we want to keep it at 2. The final ranks from the algorithm will be:
```go
map[string]int {
   "a": 0,
   "d": 0,
   "b": 1,
   "c": 2,
   "f": 3,
}
```

We can never set `c` to have rank 1 otherwise it will run in parallel with `b` which should happen prior to `c`. Hope this makes sense!

Reply:
Yeah now I agree we should set `c` to 2. However, if we start from `d` then since `c` is visited, it won't get updated to a higher rank `2`

Reply:
it will boss because of this `if` statement. 
Since each time `traverse` is called `marked` is reset. After starting from `d` we will have: 
```
ranks: {"d": 0, "c": 1, f: "2"}
```
Then we reset all marks so `c` is no longer marked. 
When we start from `a`, in this if-statement we get:
```go
if rank /* 2 */ := bfs.ranks[vtx] /* rank of "b" = 1 */ + 1; rank > bfs.ranks[neighbor ] /*  2 > 1 so update ranks["c"] */
```